### PR TITLE
tp-qemu: Add case for balloon hotplug.

### PR DESCRIPTION
**Pull request by CongSmile**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1285979

### CODE REVIEW

**Comment by CongSmile** [comment]
Windows related cmd will update in the guest-os.cfg.

File: qemu/tests/pci_hotplug.py [hunk 1]
```diff
@@ -92,6 +92,16 @@ def verify_supported_device(dev):
...
+    def device_add_balloon(pci_num, queues=1):
+        device_id = pci_type + "-" + utils_misc.generate_random_id()
+        pci_info.append([device_id, device_id])
+
+        pci_model = params.get("pci_model")
+        verify_supported_device(pci_model)
+        pci_add_cmd = "device_add id=%s,driver=%s" % (pci_info[pci_num][1],
+                                                      pci_model)
+        return device_add(pci_num, pci_add_cmd)
+
```
Comment (suqinhuang) [inline qemu/tests/pci_hotplug.py]: Hi coli,

suggest to separate  balloon hotplug, pci_hotplug will be longer and longer,
or you can put it in qemu.py like memory hotplug

**Comment by CongSmile** [comment]
@suqinhuang 

Ok, will use qdevice to rewrite it.

Thanks.

**Comment by CongSmile** [comment]
@suqinhuang and @xutian 

I have updated it. Could you help review this PR?

Thanks.

File: qemu/tests/balloon_hotplug.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    pause = float(params.get("virtio_balloon_pause", 3.0))
+    idx = 0
+    err = ""
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    error_context.context("Hotplug and unplug balloon device in a loop",
+                          logging.info)
+    for i in xrange(int(params.get("virtio_balloon_loops", 3))):
+        new_dev = qdevices.QDevice("virtio-balloon-pci",
```
Comment (spcui) [inline qemu/tests/balloon_hotplug.py]: need to vm.devices.set_dirty() above this line?

Comment (CongSmile) [inline qemu/tests/balloon_hotplug.py]: Could you help explain why need add this func()?

    def set_dirty(self):
        """ Increase VM dirtiness (not synchronized with VM) """
        if self.__state >= 0:
            self.__state += 1 
        else:
            self.__state = 1

**Comment by CongSmile** [comment]
@spcui 

I have added set_dirty() and set_clean() for the case, could you help review it again?

Thanks.

**Comment by spcui** [comment]
LGTM
Acked-by: scui@redhat.com

File: qemu/tests/balloon_hotplug.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        if ver_out is False:
+            err += ("\nDevice is not in qtree %ss after hotplug:\n%s"
+                    % (pause, vm.monitor.info("qtree")))
+
+        error_context.context("Check whether balloon device work after hotplug",
+                              logging.info)
+        memory_balloon()
+
+        # Unplug
+        out = new_dev.unplug(vm.monitor)
```
Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: could you pls add step descript to error_context

File: qemu/tests/balloon_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    pause = float(params.get("virtio_balloon_pause", 3.0))
+    idx = 0
+    err = ""
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    error_context.context("Hotplug and unplug balloon device in a loop",
+                          logging.info)
+    for i in xrange(int(params.get("virtio_balloon_loops", 3))):
```
Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: how about "balloon_repeat" instead of "virtio_balloon_loops"

File: qemu/tests/balloon_hotplug.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        vm.devices.set_dirty()
+        new_dev = qdevices.QDevice("virtio-balloon-pci",
+                                   {'id': 'balloon%d' % idx},
+                                   parent_bus={'aobject': 'pci.0'})
+
+        # Hotplug
+        out = new_dev.hotplug(vm.monitor)
```
Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: Could you pls add step descript to error_context

**Comment by CongSmile** [comment]
@suqinhuang 

I have updated the code according to your comments, could you help review it again?

Thanks.

File: qemu/tests/balloon_hotplug.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    def memory_balloon():
+        if params['os_type'] == 'windows':
+            balloon_test = balloon_check.BallooningTestWin(test, params, env)
+        else:
+            balloon_test = balloon_check.BallooningTestLinux(test, params, env)
+        for tag in params.objects('test_tags'):
+            params_tag = params.object_params(tag)
+            error_context.context("Running %s test" % tag, logging.info)
+            balloon_type = params_tag['balloon_type']
+            min_sz, max_sz = balloon_test.get_memory_boundary(balloon_type)
```
Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: (monitor qmp1)    actual: 1073741824
Sending command: cat /proc/meminfo |grep MemFree
list index out of range
Context: Running evict test
Sending command: cat /proc/meminfo |grep MemFree
list index out of range

need to raise error

File: qemu/tests/balloon_hotplug.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        if out:
+            err += "\nUnplug monitor output: " + out
+        # Pause
+        time.sleep(pause)
+        ver_out = new_dev.verify_unplug(out, vm.monitor)
+        if ver_out is False:
+            err += ("\nDevice is still in qtree %ss after unplug:\n%s"
+                    % (pause, vm.monitor.info("qtree")))
+
+        if err != "":
```
Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: how about if error is not None

File: qemu/tests/balloon_hotplug.py (hunk 6 above)
Comment (CongSmile) [inline qemu/tests/balloon_hotplug.py]: Oh, sorry, I forgot to push another PR.

Comment (CongSmile) [inline qemu/tests/balloon_hotplug.py]: Oh, sorry, I missed one PR, please help apply https://github.com/avocado-framework/avocado-vt/pull/314 .

Thanks.

File: qemu/tests/balloon_hotplug.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+    error_context.context("Hotplug and unplug balloon device in a loop",
+                          logging.info)
+    for i in xrange(int(params.get("balloon_repeats", 3))):
+        vm.devices.set_dirty()
+        new_dev = qdevices.QDevice("virtio-balloon-pci",
+                                   {'id': 'balloon%d' % idx},
+                                   parent_bus={'aobject': 'pci.0'})
+
+        error_context.context("Hotplug balloon device for %s times" % i+1,
```
Comment (qunfang) [inline qemu/tests/balloon_hotplug.py]: Hi, Cong
Why there is a "%s" in line 54, instead of "%d"?  Same question for line70.

Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: TypeError: cannot concatenate 'str' and 'int' objects

File: qemu/tests/balloon_hotplug.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        ver_out = new_dev.verify_unplug(out, vm.monitor)
+        if ver_out is False:
+            err += ("\nDevice is still in qtree %ss after unplug:\n%s"
+                    % (pause, vm.monitor.info("qtree")))
+
+        if err != "":
+            logging.error(vm.monitor.info("qtree"))
+            raise error_context.TestFail("Error occurred while hotpluging "
+                                         "virtio-pci. Iteration %s, monitor "
+                                         "output:%s" % (i, err))
```
Comment (qunfang) [inline qemu/tests/balloon_hotplug.py]: When the test failed, the log does not displayed completely:

  File "/root/test_home/qzhang/avocado-dir/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/balloon_hotplug.py", line 84, in run
    raise error_context.TestFail("Error occurred while hotpluging "
AttributeError: 'module' object has no attribute 'TestFail'

Traceback (most recent call last):

**Comment by qunfang** [comment]
LGTM.

Ack-ed by: qzhang@redhat.com

File: qemu/tests/balloon_hotplug.py (hunk 7 above)
Comment (CongSmile) [inline qemu/tests/balloon_hotplug.py]: err = "" has been initiated in line 43.

Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: if not error:

Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: tested, if err: works

**Comment by CongSmile** [comment]
@suqinhuang 

Updated. Could you help review it again?

Thanks.

File: qemu/tests/balloon_hotplug.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+        error_context.context("Unplug balloon device for %d times" % (i+1),
+                              logging.info)
+        out = new_dev.unplug(vm.monitor)
+        if out:
+            err += "\nUnplug monitor output: " + out
+        # Pause
+        time.sleep(pause)
+        ver_out = new_dev.verify_unplug(out, vm.monitor)
+        if ver_out is False:
```
Comment (CongSmile) [inline qemu/tests/balloon_hotplug.py]: if not ver_out:

Comment (suqinhuang) [inline qemu/tests/balloon_hotplug.py]: if not ver_out

**Comment by CongSmile** [comment]
17:23:09 INFO | PASS smp_2.4096m.repeat1.run_test.Host_RHEL.m7.u1.qcow2.virtio_blk.up.virtio_net.Win8.x86_64.1.io-github-autotest-qemu.balloon_hotplug

**Comment by suqinhuang** [comment]
@CongSmile
As we discussed, did you check why not use balloon_check directly?

**Comment by CongSmile** [comment]
@suqinhuang 
It use the balloon_check directly now.

**Comment by xutian** [comment]
Looks ok for me.  @suqinhuang  any comments on it ?

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

**Comment by xutian** [comment]
Fine, merged it. thanks @CongSmile , @qunfang  and @spcui  and @suqinhuang

**Pull request merged**